File: scratch/parse_noticias.js
const fs = require('fs');
const path = require('path');

function generateSlug(title) {
  // Simple slugifier
  return title
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

function parseNoticiaFile() {
  const noticiaPath = 'C:\\Users\\Dell\\Documents\\codeengine1.2\\noticia.md';
  const content = fs.readFileSync(noticiaPath, 'utf8');

  // 1. Parse Image URLs
  const imageRegex = /https:\/\/images\.pexels\.com\/photos\/\d+\/[^\s\n]+/g;
  const images = [];
  let imgMatch;
  while ((imgMatch = imageRegex.exec(content)) !== null) {
    images.push(imgMatch[0]);
  }
  console.log(`Found ${images.length} Pexels image URLs.`);

  // 2. Parse structured articles by splitting on "Título (PT)"
  const articleBlocks = content.split("Título (PT)\n");
  const parsedArticles = [];

  // Ignore intro block
  for (let i = 1; i < articleBlocks.length; i++) {
    const block = articleBlocks[i];
    const lines = block.split('\n').map(l => l.trim());

    const title = lines[0];

    // Find thumbnail
    let thumbUrl = "N/A";
    const thumbIdx = lines.indexOf("URL da Thumbnail (opcional)");
    if (thumbIdx !== -1 && lines[thumbIdx + 1]) {
      thumbUrl = lines[thumbIdx + 1];
    }

    if (thumbUrl === "N/A" || !thumbUrl) {
      if (i - 1 < images.length) {
        thumbUrl = images[i - 1];
      }
    }

    // Find tags
    let tags = [];
    const tagsIdx = lines.indexOf("Tags (separadas por vírgula)");
    if (tagsIdx !== -1 && lines[tagsIdx + 1]) {
      tags = lines[tagsIdx + 1].split(',').map(t => t.trim()).filter(t => t);
    }

    // Find summary
    let summary = "";
    const sumIdx = lines.indexOf("Resumo (PT)");
    if (sumIdx !== -1 && lines[sumIdx + 1]) {
      summary = lines[sumIdx + 1];
    }

    // Find content
    let contentText = "";
    const contentMarker = "Conteúdo (PT)";
    const markerIndex = block.indexOf(contentMarker);
    if (markerIndex !== -1) {
      contentText = block.substring(markerIndex + contentMarker.length).trim();
    }

    // Extract conclusion if present
    let conclusionText = "";
    const conclMarker = "Conclusão: A Transferência de Poder";
    if (contentText.includes(conclMarker)) {
      const parts = contentText.split(conclMarker);
      contentText = parts[0].trim();
      conclusionText = parts[1].trim();
    }

    // Categories available: AI, Automação, SaaS, Programação, Produtividade, Inovação
    let category = 'Inovação';
    const hasAI = tags.some(t => /IA|openai|claude|inteligência|intelligence/i.test(t));
    const hasAutomation = tags.some(t => /automação|automation|copilot|builder/i.test(t));
    const hasSaaS = tags.some(t => /saas|startups|mercado|business/i.test(t));
    const hasProgramming = tags.some(t => /programação|programming|desenvolvedor|coding/i.test(t));
    const hasProductivity = tags.some(t => /produtividade|productivity/i.test(t));

    if (hasAI) category = 'AI';
    else if (hasAutomation) category = 'Automação';
    else if (hasSaaS) category = 'SaaS';
    else if (hasProgramming) category = 'Programação';
    else if (hasProductivity) category = 'Produtividade';

    parsedArticles.push({
      title,
      slug: generateSlug(title),
      excerpt: summary,
      content: contentText,
      thumbnail_url: thumbUrl,
      tags,
      category
    });

    if (conclusionText) {
      const conclTitle = "Conclusão: A Transferência de Poder";
      const conclThumb = images[29] || images[images.length - 1];
      parsedArticles.push({
        title: conclTitle,
        slug: generateSlug(conclTitle),
        excerpt: "Uma análise profunda sobre a transferência de poder e infraestrutura na era das IAs autônomas.",
        content: conclusionText,
        thumbnail_url: conclThumb,
        tags: ["IA", "Infraestrutura", "Geopolítica", "Tendências", "Destaque"],
        category: 'Inovação'
      });
    }
  }

  console.log(`Successfully parsed ${parsedArticles.length} articles.`);
  return parsedArticles;
}

function generateSql(articles) {
  const sqlLines = [];
  sqlLines.push("-- ==========================================================");
  sqlLines.push("-- MIGRATION: SEED NEWS DATA FROM NOTICIA.MD");
  sqlLines.push("-- ==========================================================");
  sqlLines.push("");
  sqlLines.push("BEGIN;");
  sqlLines.push("");
  sqlLines.push("-- Clear existing news to prevent conflicts");
  sqlLines.push("DELETE FROM news;");
  sqlLines.push("");

  const baseTime = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000); // 30 days ago

  articles.forEach((art, idx) => {
    const titleEsc = art.title.replace(/'/g, "''");
    const slugEsc = art.slug.replace(/'/g, "''");
    const excerptEsc = art.excerpt.replace(/'/g, "''");
    const contentEsc = art.content.replace(/'/g, "''");
    const thumbEsc = art.thumbnail_url.replace(/'/g, "''");
    const categoryEsc = art.category.replace(/'/g, "''");

    const tagsFormatted = art.tags.map(t => `'${t.replace(/'/g, "''")}'`).join(', ');
    const tagsArray = tagsFormatted ? `ARRAY[${tagsFormatted}]` : 'NULL';

    // Calculate dates incrementally
    const pubDate = new Date(baseTime.getTime() + idx * 6 * 60 * 60 * 1000); // add 6 hours per article
    const pubDateStr = pubDate.toISOString();

    sqlLines.push(`-- Artigo ${idx + 1}: ${art.title}`);
    sqlLines.push("INSERT INTO news (");
    sqlLines.push("  title, slug, excerpt, content, thumbnail_url, category, tags, author, status, published_at, created_at, updated_at");
    sqlLines.push(") VALUES (");
    sqlLines.push(`  '${titleEsc}',`);
    sqlLines.push(`  '${slugEsc}',`);
    sqlLines.push(`  '${excerptEsc}',`);
    sqlLines.push(`  '${contentEsc}',`);
    sqlLines.push(`  '${thumbEsc}',`);
    sqlLines.push(`  '${categoryEsc}',`);
    sqlLines.push(`  ${tagsArray},`);
    sqlLines.push("  'Fernando JR',");
    sqlLines.push("  'published',");
    sqlLines.push(`  '${pubDateStr}',`);
    sqlLines.push(`  '${pubDateStr}',`);
    sqlLines.push(`  '${pubDateStr}'`);
    sqlLines.push(") ON CONFLICT (slug) DO UPDATE SET");
    sqlLines.push("  title = EXCLUDED.title,");
    sqlLines.push("  excerpt = EXCLUDED.excerpt,");
    sqlLines.push("  content = EXCLUDED.content,");
    sqlLines.push("  thumbnail_url = EXCLUDED.thumbnail_url,");
    sqlLines.push("  category = EXCLUDED.category,");
    sqlLines.push("  tags = EXCLUDED.tags,");
    sqlLines.push("  published_at = EXCLUDED.published_at,");
    sqlLines.push("  updated_at = NOW();");
    sqlLines.push("");
  });

  sqlLines.push("COMMIT;");
  sqlLines.push("");
  sqlLines.push("SELECT '✅ Seed completed successfully! 30 articles inserted.' AS status;");

  const sqlPath = 'C:\\Users\\Dell\\Documents\\codeengine1.2\\scratch\\insert_noticias.sql';
  fs.writeFileSync(sqlPath, sqlLines.join('\n'), 'utf8');
  console.log(`SQL script generated at: ${sqlPath}`);
}

const articles = parseNoticiaFile();
generateSql(articles);

File: scratch/parse_noticias.py
import re
import json

def generate_slug(title):
    # Normalize and convert to clean slug
    import unicodedata
    title = unicodedata.normalize('NFD', title).encode('ascii', 'ignore').decode('utf-8')
    title = title.lower()
    title = re.sub(r'[^a-z0-9]+', '-', title)
    title = re.sub(r'^-+|-+$', '', title)
    return title

def parse_noticia_file():
    with open(r'C:\Users\Dell\Documents\codeengine1.2\noticia.md', 'r', encoding='utf-8') as f:
        content = f.read()

    # 1. Parse Image URLs
    image_pattern = r'https://images\.pexels\.com/photos/\d+/[^\s\n]+'
    images = re.findall(image_pattern, content)
    print(f"Found {len(images)} Pexels image URLs.")

    # 2. Parse structured articles
    # Each article starts with "Título (PT)"
    article_blocks = content.split("Título (PT)\n")
    # The first split part contains the images list
    intro_part = article_blocks[0]
    article_blocks = article_blocks[1:]
    
    parsed_articles = []
    
    for i, block in enumerate(article_blocks):
        lines = [line.strip() for line in block.split('\n')]
        
        # Title is the first line
        title = lines[0]
        
        # Find Thumbnail URL (optional)
        thumb_url = "N/A"
        try:
            thumb_idx = lines.index("URL da Thumbnail (opcional)")
            thumb_url = lines[thumb_idx + 1]
        except ValueError:
            pass
            
        # If thumbnail is N/A or empty, use the mapped image from the list
        if thumb_url == "N/A" or not thumb_url:
            if i < len(images):
                thumb_url = images[i]
                
        # Find Tags
        tags = []
        try:
            tags_idx = lines.index("Tags (separadas por vírgula)")
            tags_str = lines[tags_idx + 1]
            tags = [t.strip() for t in tags_str.split(',') if t.strip()]
        except ValueError:
            pass
            
        # Find Summary/Excerpt
        summary = ""
        try:
            sum_idx = lines.index("Resumo (PT)")
            summary = lines[sum_idx + 1]
        except ValueError:
            pass
            
        # Find Content
        # Content starts after "Conteúdo (PT)" and runs until "Conclusão: A Transferência de Poder" or the end of the block
        content_text = ""
        try:
            content_idx = block.index("Conteúdo (PT)") + len("Conteúdo (PT)")
            content_text = block[content_idx:].strip()
        except ValueError:
            pass
            
        # For the 29th article, extract the "Conclusão" section if present
        conclusion_text = ""
        if "Conclusão: A Transferência de Poder" in content_text:
            parts = content_text.split("Conclusão: A Transferência de Poder")
            content_text = parts[0].strip()
            conclusion_text = parts[1].strip()
            
        # Append parsed article
        parsed_articles.append({
            'title': title,
            'slug': generate_slug(title),
            'excerpt': summary,
            'content': content_text,
            'thumbnail_url': thumb_url,
            'tags': tags,
            'category': 'AI' if 'IA' in tags or 'OpenAI' in tags or 'Claude' in tags or 'Flux' in tags else 'Inovação'
        })
        
        # If we extracted the conclusion, it becomes the 30th article!
        if conclusion_text:
            concl_title = "Conclusão: A Transferência de Poder"
            concl_thumb = images[29] if len(images) > 29 else images[-1]
            parsed_articles.append({
                'title': concl_title,
                'slug': generate_slug(concl_title),
                'excerpt': "Uma análise profunda sobre a transferência de poder e infraestrutura na era das IAs autônomas.",
                'content': conclusion_text,
                'thumbnail_url': concl_thumb,
                'tags': ["IA", "Infraestrutura", "Geopolítica", "Tendências", "Destaque"],
                'category': 'Inovação'
            })

    print(f"Successfully parsed {len(parsed_articles)} articles.")
    return parsed_articles

def generate_sql(articles):
    sql_lines = []
    sql_lines.append("-- ==========================================================")
    sql_lines.append("-- MIGRATION: SEED NEWS DATA FROM NOTICIA.MD")
    sql_lines.append("-- ==========================================================")
    sql_lines.append("")
    sql_lines.append("BEGIN;")
    sql_lines.append("")
    sql_lines.append("-- Optional: clean existing news to prevent slug conflicts")
    sql_lines.append("DELETE FROM news;")
    sql_lines.append("")
    
    # We want these news to appear first, so we can set their published_at
    # timestamps in descending order, starting from today going backwards,
    # or starting from a few days ago going forward.
    # To make them appear as the latest, we will give them a clean timestamp sequence!
    import datetime
    base_time = datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(days=30)
    
    for i, art in enumerate(articles):
        # Escape quotes in text
        title_esc = art['title'].replace("'", "''")
        slug_esc = art['slug'].replace("'", "''")
        excerpt_esc = art['excerpt'].replace("'", "''")
        content_esc = art['content'].replace("'", "''")
        thumb_esc = art['thumbnail_url'].replace("'", "''")
        category_esc = art['category'].replace("'", "''")
        
        # Format tags as PostgreSQL array literal, e.g., ARRAY['IA', 'OpenAI']
        tags_formatted = ", ".join([f"'{t.replace(\"'\", \"''\")}'" for t in art['tags']])
        tags_array = f"ARRAY[{tags_formatted}]" if tags_formatted else "NULL"
        
        # Distribute published dates so they appear in correct chronological order
        # Let's add 1 hour or 1 day incrementally so that the latest articles are on top
        pub_date = base_time + datetime.timedelta(hours=i*6)
        pub_date_str = pub_date.strftime('%Y-%m-%d %H:%M:%S%z')
        # Format timezone correctly for PG (e.g. +0000 -> +00:00)
        if pub_date_str.endswith('00'):
            pub_date_str = pub_date_str[:-2] + ':' + pub_date_str[-2:]
            
        sql_lines.append(f"-- Artigo {i+1}: {art['title']}")
        sql_lines.append("INSERT INTO news (")
        sql_lines.append("  title, slug, excerpt, content, thumbnail_url, category, tags, author, status, published_at, created_at, updated_at")
        sql_lines.append(") VALUES (")
        sql_lines.append(f"  '{title_esc}',")
        sql_lines.append(f"  '{slug_esc}',")
        sql_lines.append(f"  '{excerpt_esc}',")
        sql_lines.append(f"  '{content_esc}',")
        sql_lines.append(f"  '{thumb_esc}',")
        sql_lines.append(f"  '{category_esc}',")
        sql_lines.append(f"  {tags_array},")
        sql_lines.append("  'Fernando JR',")
        sql_lines.append("  'published',")
        sql_lines.append(f"  '{pub_date_str}',")
        sql_lines.append(f"  '{pub_date_str}',")
        sql_lines.append(f"  '{pub_date_str}'")
        sql_lines.append(") ON CONFLICT (slug) DO UPDATE SET")
        sql_lines.append("  title = EXCLUDED.title,")
        sql_lines.append("  excerpt = EXCLUDED.excerpt,")
        sql_lines.append("  content = EXCLUDED.content,")
        sql_lines.append("  thumbnail_url = EXCLUDED.thumbnail_url,")
        sql_lines.append("  category = EXCLUDED.category,")
        sql_lines.append("  tags = EXCLUDED.tags,")
        sql_lines.append("  published_at = EXCLUDED.published_at,")
        sql_lines.append("  updated_at = NOW();")
        sql_lines.append("")
        
    sql_lines.append("COMMIT;")
    sql_lines.append("")
    sql_lines.append("SELECT '✅ Seed completed successfully! 30 articles inserted.' AS status;")
    
    with open(r'C:\Users\Dell\Documents\codeengine1.2\scratch\insert_noticias.sql', 'w', encoding='utf-8') as f:
        f.write("\n".join(sql_lines))
    print("SQL script generated at: C:\\Users\\Dell\\Documents\\codeengine1.2\\scratch\\insert_noticias.sql")

if __name__ == "__main__":
    articles = parse_noticia_file()
    generate_sql(articles)
